crates/bevy_ecs/src/relationship/mod.rs

```diff
@@ -185,38 +231,70 @@ pub trait Relationship: Component + Sized {
                 }
             }
         }
-        let target_entity = world.entity(entity).get::<Self>().unwrap().get();
-        if let Ok(mut target_entity_mut) = world.get_entity_mut(target_entity) {
-            if let Some(mut relationship_target) =
-                target_entity_mut.get_mut::<Self::RelationshipTarget>()
-            {
-                relationship_target.collection_mut_risky().remove(entity);
-                if relationship_target.len() == 0 {
-                    let command = |mut entity: EntityWorldMut| {
-                        // this "remove" operation must check emptiness because in the event that an identical
-                        // relationship is inserted on top, this despawn would result in the removal of that identical
-                        // relationship ... not what we want!
-                        if entity
-                            .get::<Self::RelationshipTarget>()
-                            .is_some_and(RelationshipTarget::is_empty)
-                        {
-                            entity.remove::<Self::RelationshipTarget>();
-                        }
-                    };
+        // We need to access two entities simultaneously (one now, and another later),
+        // which isn't possible via safe APIs currently.
+        let world_cell = world.as_unsafe_world_cell();
+        // SAFETY: We're only accessing this entity's `Relationship` component currently,
+        // which doesn't violate any safety invariants.
+        let target_entities = unsafe { world_cell.get_entity(entity).unwrap().get::<Self>() }
+            .unwrap()
+            .get();
+        for target_entity in target_entities.iter() {
+            if target_entity == entity {
+                // Self-referential relationships are cleaned up by the `on_insert`
+                // hook, but we may still encounter them when they're being removed.
+                // We skip them here to avoid potentially accessing the same entity
+                // simultaneously, which could violate rust safety invariants if
+                // the `Relationship` and `RelationshipTarget` are the same component
+                // (which is technically possible for many-to-many relationships).
+                continue;
+            }
 
-                    world
-                        .commands()
+            if let Ok(target_entity_cell) = world_cell.get_entity(target_entity) {
```

Comment:
So if the source_entity has a Relationship with 10 entities and we add an 11th; we will first apply the remove on all 10 entities in `on_replace` (potentially queuing up 10 remove commands), and then in `on_insert` we queue up 11 commands to add the entity?

It's probably fine since relationship modifications are usually rare, but maybe this should be documented? (updating the relationship 1 by 1 is a O(n2) op)

Maybe in the future we could have an improved ON_REPLACE observer that gives you both the existing component value and the new component value. Theoretically it should be doable since both are accessible here: https://github.com/bevyengine/bevy/blob/45c997c45ff740a918fb74d5013d00b0b846b571/crates/bevy_ecs/src/bundle/insert.rs#L175

Reply:
Maybe we could have a flag (like LINKED_SPAWN); if present we would use a diffing approach to compare the old and the new values.
The diffing would require changes to the ON_REPLACE observer; or adding another component like `CachedRelationship` that holds the previous values of the Relationship component.

This could be nice in particular for RelationshipCollections like HashSet or IndexSet where you can efficiently get the diff.

Out of scope for this PR though